### PR TITLE
[shardformer] update llama2/opt finetune example and fix llama2 policy

### PR DESCRIPTION
**Pull request by flybird11111**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

#4582 
#4643 
#4644 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

update llama2/opt finetune example and shardformer update to llama2

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: tests/kit/model_zoo/transformers/llama.py [hunk 1]
```diff
@@ -52,6 +52,9 @@ def data_gen_for_casual_lm():
                          max_position_embeddings=128,
                          num_labels=16)
 
+    if hasattr(config, "pad_token_id"):
+        config.pad_token_id = 2
```
Comment (ver217) [inline tests/kit/model_zoo/transformers/llama.py:56]: Why do so?

Comment (flybird11111) [reply tests/kit/model_zoo/transformers/llama.py:56]: llama2config has added this parameter. In the LlamaForSequenceClassification model's forward function, there is such a condition check: `        if self.config.pad_token_id is None and batch_size != 1:
            raise ValueError("Cannot handle batch sizes > 1 if no padding token is defined.")`

File: examples/language/llama2/finetune.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    train_dataloader = data_builder.train_dataloader()
+    test_dataloader = data_builder.test_dataloader()
+
+    # ====================================
+    # Prepare model, optimizer
+    # ====================================
+
+    cfg = AutoConfig.from_pretrained(args.model_path)
+
+    model = LlamaForCausalLM.from_pretrained(args.model_path, config=cfg).cuda()
```
Comment (ver217) [inline examples/language/llama2/finetune.py:149]: Add lazy init option and demonstrate when to use it by comment

File: examples/language/llama2/finetune.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+                                                                  optimizer,
+                                                                  criterion=_criterion,
+                                                                  lr_scheduler=lr_scheduler)
+
+    # ==============================
+    # Train model
+    # ==============================
+    for epoch in range(NUM_EPOCHS):
+        train_epoch(epoch, model, optimizer, _criterion, lr_scheduler, train_dataloader, booster, coordinator)
+
```
Comment (ver217) [inline examples/language/llama2/finetune.py:193]: Add saving

File: examples/language/llama2/finetune.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        "stage_manager") and booster.plugin.stage_manager is not None and booster.plugin.stage_manager.is_last_stage()
+    with tqdm(train_dataloader,
+              desc=f'Epoch [{epoch + 1}/{NUM_EPOCHS}]',
+              disable=not (coordinator.is_master() or is_pp_last_stage)) as pbar:
+        for batch in pbar:
+            # Forward pass
+            batch = move_to_cuda(batch)
+            if hasattr(booster.plugin, "stage_manager") and booster.plugin.stage_manager is not None:
+                #TODO pass train_dataloader to execute_pipeline directly
+                batch = iter([batch])
```
Comment (ver217) [inline examples/language/llama2/finetune.py:56]: DON'T pass batch to pipeline. See https://github.com/hpcaitech/ColossalAI/pull/4641

File: examples/language/opt/opt_train_demo.py [hunk 5]
```diff
@@ -11,43 +11,65 @@
...
 
             # Forward
             optimizer.zero_grad()
             batch = move_to_cuda(batch, torch.cuda.current_device())
 
-            outputs = model(use_cache=False, **batch)
-            loss = outputs['loss']
+            if hasattr(booster.plugin, "stage_manager") and booster.plugin.stage_manager is not None:
+                #TODO pass train_dataloader to execute_pipeline directly
+                batch = iter([batch])
```
Comment (ver217) [inline examples/language/opt/opt_train_demo.py:51]: same problem

File: examples/language/llama2/finetune.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+    is_pp_last_stage = hasattr(
+        booster.plugin,
+        "stage_manager") and booster.plugin.stage_manager is not None and booster.plugin.stage_manager.is_last_stage()
+    with tqdm(train_dataloader,
+              desc=f'Epoch [{epoch + 1}/{NUM_EPOCHS}]',
+              disable=not (coordinator.is_master() or is_pp_last_stage)) as pbar:
+        for batch in pbar:
+            # Forward pass
+            batch = move_to_cuda(batch)
+            if hasattr(booster.plugin, "stage_manager") and booster.plugin.stage_manager is not None:
```
Comment (ver217) [inline examples/language/llama2/finetune.py:54]: This judgement is weird. Stage manager is transparent to users. You should use clearer api.

Comment (ver217) [reply examples/language/llama2/finetune.py:54]: FYI https://github.com/hpcaitech/ColossalAI/pull/4641/files#diff-a738d0ca17d7f18ecd94e41a76e0327668b997f5c88d18957cb0274a8b256f9bR32

File: tests/kit/model_zoo/transformers/llama.py (hunk 1 above)
Comment (ver217) [reply tests/kit/model_zoo/transformers/llama.py:56]: pad_token is set to unk_token in llama2

File: examples/language/llama2/finetune.py (hunk 2 above)
Comment (flybird11111) [reply examples/language/llama2/finetune.py]: updated.

File: examples/language/llama2/finetune.py (hunk 3 above)
Comment (flybird11111) [reply examples/language/llama2/finetune.py:202]: updated.

File: examples/language/llama2/finetune.py (hunk 4 above)
Comment (flybird11111) [reply examples/language/llama2/finetune.py]: updated.

File: examples/language/llama2/finetune.py (hunk 6 above)
Comment (flybird11111) [reply examples/language/llama2/finetune.py]: updated

File: examples/language/llama2/finetune.py [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+    # ====================================
+    # Prepare model, optimizer
+    # ====================================
+
+    cfg = AutoConfig.from_pretrained(args.model_path)
+
+    if args.use_lazy_init:
+        args.use_lazy_init = False
+        warnings.warn("lazy init is not compatible with from_pretrained now")
```
Comment (ver217) [inline examples/language/llama2/finetune.py:154]: Lazy init can be used by initializing from config and load model after `boost`

Comment (ver217) [reply examples/language/llama2/finetune.py:154]: You shoud add comments here to demonstrate in which case (model and plugin) we should use lazy init.

**Comment by ver217** [comment]
@TongLi3701 Hi, could you please review this PR? If you find any inconvenient parts or have any requirements, please comment this PR.

File: examples/language/bert/finetune.py [hunk 8]
```diff
@@ -59,17 +59,21 @@ def evaluate_model(
...
             labels = batch["labels"]
             batch_size = batch["input_ids"].shape[0]
-            if hasattr(booster.plugin, "stage_manager") and booster.plugin.stage_manager is not None:
+            if use_pipeline:
                 pg_mesh = booster.plugin.pg_mesh
                 pp_group = booster.plugin.pp_group
                 current_pp_group_ranks = pg_mesh.get_ranks_in_group(pp_group)
                 current_rank = dist.get_rank()
-                #TODO pass dataloader to execute_pipeline directly
+                # Can't pass dataloader to execute_pipeline directly, Because we need the actual batch size from batch to broadcast output.
```
Comment (ver217) [inline examples/language/bert/finetune.py:76]: Batch size can be got from predicts

File: examples/language/bert/finetune.py [hunk 9]
```diff
@@ -138,31 +142,33 @@ def evaluate_subset(dataloader: DataLoader):
...
+    is_pp_last_stage = use_pipeline and booster.plugin.stage_manager.is_last_stage()
+    total_step = len(train_dataloader)
+
     model.train()
-    is_pp_last_stage = hasattr(
-        booster.plugin,
-        "stage_manager") and booster.plugin.stage_manager is not None and booster.plugin.stage_manager.is_last_stage()
-    with tqdm(train_dataloader,
+    optimizer.zero_grad()
+    train_dataloader = iter(train_dataloader)
```
Comment (ver217) [inline examples/language/bert/finetune.py:151]: Name it `train_data_iter` would be better.

**Review by ver217 — APPROVED** [review]

**Review by FrankLeeeee — APPROVED** [review]

**Comment by TongLi3701** [comment]
> @TongLi3701 Hi, could you please review this PR? If you find any inconvenient parts or have any requirements, please comment this PR.

Sure, I will have a look!

**Comment by flybird11111** [comment]
> > @TongLi3701 Hi, could you please review this PR? If you find any inconvenient parts or have any requirements, please comment this PR.
> 
> Sure, I will have a look!

hihi, llama2 fine-tune example moves to this pr. https://github.com/hpcaitech/ColossalAI/pull/4673
